Core/src/org/sleuthkit/autopsy/casemodule/services/TagsManager.java

```diff
@@ -166,21 +222,31 @@ public synchronized TagName addTagName(String displayName, String description, T
         if (null == caseDb) {
             throw new TskCoreException("Tags manager has been closed");
         }
-        lazyLoadExistingTagNames();
-        if (uniqueTagNames.containsKey(displayName)) {
-            throw new TagNameAlreadyExistsException();
-        }
 
+        lazyLoadExistingTagNames();
+        
         /*
-         * Add the tag name to the case.
+         * It is possible user is trying to add back a tag after having deleted
+         * it. It is also possible a user tag name was never added to the
+         * database.
          */
+        if (uniqueTagNames.containsKey(displayName)) {
+            if (uniqueTagNames.get(displayName) != null) {
+                TagName existingTagName = uniqueTagNames.get(displayName);
+                long count = getContentTagsCountByTagName(existingTagName) + getBlackboardArtifactTagsCountByTagName(existingTagName);
+                if (count == 0) {
+                    addNewTagNameToTagsSettings(existingTagName);
+                    return existingTagName;
+                } else {
+                    throw new TagNameAlreadyExistsException();
+                }
+            }
+        }
+        
```

Comment:
I don't think throwing the TagNameAlreadyExistsException is currently useful given that there is a race condition inherent in the creation of tag names for multi-user cases. I think we should check if we have a matching TagName object in memory. If not, we should try to add it to the database. If an exception is thrown by SleuthkitCase.addTagName, it may be due to a database issue, or it may be due to an attempt to duplicate a tag added by another user. To allow for the latter, we should "refresh" the tags in memory from the database and see if we can get a matching tag name TagName that way. If the call to SleuthkitCase.getAllTagNames for the "refresh" throws, the exception should propagate.

We will keep TagNameAlreadyExistsException in the exception specification because to change it would be a public API change, and we will probably need it anyway, if we get to the point where we use the description and color columns in the tag names table.